### PR TITLE
fuzz: Add PartiallyDownloadedBlock target

### PR DESCRIPTION
**Pull request by dergoegge**
This PR adds a fuzz target for `PartiallyDownloadedBlock`, which we currently do not have any coverage for.

### CODE REVIEW

**Review by mzumsande — COMMENTED** [review]
Concept ACK, only minor comments

File: src/test/fuzz/partially_downloaded_block.cpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+    auto init_status{pdb.InitData(cmpctblock, extra_txn)};
+
+    std::vector<CTransactionRef> missing;
+    for (size_t i = 0; i < cmpctblock.BlockTxCount(); i++) {
+        bool skip{fuzzed_data_provider.ConsumeBool()};
+        assert(init_status != READ_STATUS_OK ||
+               (available.count(i) == 1) == pdb.IsTxAvailable(i));
+        if (!pdb.IsTxAvailable(i) && !skip && i < block->vtx.size()) {
+            missing.push_back(block->vtx[i]);
```
[the remark this answers is not in the document]
Comment (dergoegge) [reply src/test/fuzz/partially_downloaded_block.cpp:100]: Good idea, added an extra commit for this

File: src/test/fuzz/partially_downloaded_block.cpp [hunk 2]
```diff
@@ -96,6 +99,8 @@ FUZZ_TARGET_INIT(partially_downloaded_block, initialize_pdb)
         if (!pdb.IsTxAvailable(i) && !skip && i < block->vtx.size()) {
             missing.push_back(block->vtx[i]);
         }
+
+        skipped_missing = skipped_missing || (!pdb.IsTxAvailable(i) && skip);
```
Comment (MarcoFalke) [inline src/test/fuzz/partially_downloaded_block.cpp:103]: nit: Use `|=`?

**Review by maflcko — APPROVED** [review]
Thanks for working on improved fuzz targets!

I've left some nano-nits that can be ignored. Please let us know if you want this merged or want to address the nits.

review ACK b350ffd07c04d3b499f5960973d664d699341ec1 🎧

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK b350ffd07c04d3b499f5960973d664d699341ec1 🎧
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgq+Av/em84tR5X4rXmHJSeQPxVnzZUNBTZYXwnjL+A3EbtpxWP+ne/jGTdkj43
URmm15WNSbKeLKTR3w+FbdcgNbGuyfRlx6tJsvKNFbAMZtEvR4Ezs7WxZmQKdLAa
DHsp3Tpfe+axMgu1TGpx1JZ2+wC+6pDYDqMvAh2yLkN1NHUrHEToVwIN2r4z2+Nl
0adRB0cxh7WF9VGN4twevz5AAoIEXMhlmDXEGJO8ljFyBbkxmTDtYcYzgRxdQVig
Ok9JoNuFxxVQeBTC1TeloGCqXAapwS+ZVQul0NpEsWejD29jGGi+/kw3qOXq3HjI
Kmb/K/6zNMNUgptln5bFSWnwMHAkhbKlor7a2bDPHcpO6Fiti5EubKodw3+1i0nZ
Lx5y3Jgr1ae41s/hAuxGF8MtfyBHmBhl5EkoVLnHKLE/mYmIg2C2v/TwN6xmwpuh
zXhIyon2Ht2p8pi0nI197Mp6av+Ify3/EC0UgfC+yyVHhTAugq9DIF833Y8zXrqR
VeWb6Kg0
=ay5H
-----END PGP SIGNATURE-----
```


</details>

File: src/test/fuzz/partially_downloaded_block.cpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        // compact block (i.e. IsTxAvailable(i) == true) implies that we marked
+        // that transaction as available above (i.e. available.count(i) > 0).
+        // The reverse is not true, due to possible compact block short id
+        // collisions (i.e. available.count(i) > 0 does not imply
+        // IsTxAvailable(i) == true).
+        assert(init_status != READ_STATUS_OK ||
+               !pdb.IsTxAvailable(i) || available.count(i) > 0);
+
+        bool skip{fuzzed_data_provider.ConsumeBool()};
+        if (!pdb.IsTxAvailable(i) && !skip && i < block->vtx.size()) {
```
[the remark this answers is not in the document]
Comment (dergoegge) [reply src/test/fuzz/partially_downloaded_block.cpp]: Yea this is not needed, i suspect I this is a remnant of an earlier version of the target.

File: src/test/fuzz/partially_downloaded_block.cpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+void initialize_pdb()
+{
+    static const auto testing_setup = MakeNoLogFileContext<const TestingSetup>();
+    g_setup = testing_setup.get();
+}
+
+PartiallyDownloadedBlock::CheckBlockFn FuzzedCheckBlock(std::optional<BlockValidationResult> result)
+{
+    return [result](const CBlock& block, BlockValidationState& state, const Consensus::Params& params, bool check_pow, bool check_merkle_root) {
```
[the remark this answers is not in the document]
Comment (dergoegge) [reply src/test/fuzz/partially_downloaded_block.cpp]: Dropped the unused names

**Comment by dergoegge** [comment]
@MarcoFalke addressed all your comments. CI failures seem unrelated (restarted the failing jobs).

**Review by mzumsande — COMMENTED** [review]
Code Review ACK a1c36275b5a27ae685f49ff02dabff0adbf51aa1

File: src/test/fuzz/partially_downloaded_block.cpp [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+             BlockValidationResult::BLOCK_CONSENSUS,
+             BlockValidationResult::BLOCK_RECENT_CONSENSUS_CHANGE,
+             BlockValidationResult::BLOCK_CACHED_INVALID,
+             BlockValidationResult::BLOCK_INVALID_HEADER,
+             BlockValidationResult::BLOCK_MUTATED,
+             BlockValidationResult::BLOCK_MISSING_PREV,
+             BlockValidationResult::BLOCK_INVALID_PREV,
+             BlockValidationResult::BLOCK_TIME_FUTURE,
+             BlockValidationResult::BLOCK_CHECKPOINT,
+             BlockValidationResult::BLOCK_HEADER_LOW_WORK});
```
Comment (MarcoFalke) [inline src/test/fuzz/partially_downloaded_block.cpp:121]: nit: Could move to a common header file?

**Review by maflcko — APPROVED** [review]
lgtm. Follow-ups can be done in new pulls, if there are any.


re-ACK a1c36275b5a27ae685f49ff02dabff0adbf51aa1  🎼

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK a1c36275b5a27ae685f49ff02dabff0adbf51aa1  🎼
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhUGAv8CVgILyXBbe4tlLJ6RxNL9KDFKjr7dYR+6pLtarndtMPlJs+xS/dJwHSi
6Exkiq1N54HCLzzgMZeN03rUW4FbXW1qngFiTS4SJCOliuaqtZInk4uH8O5AIlWt
igJ8Kfeues7VcwsVSIIAkfg8ZCnoj4BoOKJGWNdRVeCUfa0JqoftIHzX3yqwjmS/
7Y9CWzn7nTqqSzcel4Ry6BcqMetJsZohTtsqy8GCZfB/Ylam5TxrUgkYaBid8BO+
4sTAcy9Cj9+2JhHXeyXXVaLbPDeB3WHErHelxv8z0PgKcH1xDvvVO7SmAuKWVP6q
Lsq6n+OannMTc8V24ue5x+QQTDtTLl33C+MNSZnCG7lZ0WPW2vmHZEvyGBTsLZId
TM3uVmByZlCjOHir4Vm9hR6DupzLkjNRaoKlLSeZrrvBJuEWQ2vMZGv+5OIyHVTE
GJHsVC7djHNX1yVtqT6j8jCgqigotS5piD3hcYB3z30nBfCa1lzaZpt1p9/TkSyD
4jM6fVom
=ozo3
-----END PGP SIGNATURE-----
```


</details>

File: src/test/fuzz/partially_downloaded_block.cpp [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+            extra_txn.emplace_back(tx->GetWitnessHash(), tx);
+            available.insert(i);
+        }
+
+        if (add_to_mempool) {
+            LOCK2(cs_main, pool.cs);
+            pool.addUnchecked(ConsumeTxMemPoolEntry(fuzzed_data_provider, *tx));
+            available.insert(i);
+        }
+    }
```
Comment (MarcoFalke) [inline src/test/fuzz/partially_downloaded_block.cpp:80]: nit: Could make sense to include some extraneous txs that are not in the block in the extra txn or mempool?

File: src/test/fuzz/partially_downloaded_block.cpp [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    bool skipped_missing{false};
+    for (size_t i = 0; i < cmpctblock.BlockTxCount(); i++) {
+        // If init_status == READ_STATUS_OK then a available transaction in the
+        // compact block (i.e. IsTxAvailable(i) == true) implies that we marked
+        // that transaction as available above (i.e. available.count(i) > 0).
+        // The reverse is not true, due to possible compact block short id
+        // collisions (i.e. available.count(i) > 0 does not imply
+        // IsTxAvailable(i) == true).
+        if (init_status == READ_STATUS_OK) {
+            assert(!pdb.IsTxAvailable(i) || available.count(i) > 0);
```
Comment (MarcoFalke) [inline src/test/fuzz/partially_downloaded_block.cpp:96]: nit:

```cpp
        if (init_status == READ_STATUS_OK && pdb.IsTxAvailable(i)) {
            assert(available.count(i) > 0);
```

Also, in the comment replace "then" with "and"?

File: src/test/fuzz/partially_downloaded_block.cpp [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+        }
+
+        bool skip{fuzzed_data_provider.ConsumeBool()};
+        if (!pdb.IsTxAvailable(i) && !skip) {
+            missing.push_back(block->vtx[i]);
+        }
+
+        skipped_missing |= (!pdb.IsTxAvailable(i) && skip);
```
Comment (MarcoFalke) [inline src/test/fuzz/partially_downloaded_block.cpp:104]: nit:


```cpp
        if (!pdb.IsTxAvailable(i)) {
            if (skip) { skipped_missing = true; }
            else { missing.push_back(block->vtx[i]); }
        }

**Review by maflcko — COMMENTED** [review]
left more nits :sweat_smile:

**Pull request merged**